presto-main/src/main/java/com/facebook/presto/operator/PagesSpatialIndexSupplier.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.esri.core.geometry.ogc.OGCGeometry;
+import com.facebook.presto.Session;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.gen.JoinFilterFunctionCompiler;
+import io.airlift.slice.Slice;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import org.locationtech.jts.geom.Envelope;
+import org.locationtech.jts.index.strtree.STRtree;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiPredicate;
+
+import static com.facebook.presto.geospatial.GeometryUtils.deserialize;
+import static com.facebook.presto.operator.PagesSpatialIndex.EMPTY_INDEX;
+import static com.facebook.presto.operator.SyntheticAddress.decodePosition;
+import static com.facebook.presto.operator.SyntheticAddress.decodeSliceIndex;
+
+public class PagesSpatialIndexSupplier
+{
+    private PagesSpatialIndex index;
+
+    public PagesSpatialIndexSupplier(
+            Session session,
+            LongArrayList addresses,
+            List<Type> types,
+            List<Integer> outputChannels,
+            List<List<Block>> channels,
+            int geometryChannel,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionCompiler.JoinFilterFunctionFactory> filterFunctionFactory)
+    {
+        STRtree rtree = buildRTree(addresses, channels, geometryChannel);
+        if (rtree.isEmpty()) {
+            index = EMPTY_INDEX;
```

Comment:
Do we need this special case? Can't we just create a regular PagesRTreeIndex?

Reply:
i guess it may happen when there is lots of data on the build side, but the  geometry column is all-null. Special case allows the non-join column data to be freed earlier. I don't know, however, whether we really need this.